src/main/scala/bitboard/Bitboard.scala

```diff
@@ -0,0 +1,183 @@
+  val KING_DELTAS       = Array[Int](1, 7, 8, 9, -1, -7, -8, -9)
+  val WHITE_PAWN_DELTAS = Array[Int](7, 9)
+  val BLACK_PAWN_DELTAS = Array[Int](-7, -9)
+
+  val KNIGHT_ATTACKS     = Array.fill(64)(0L)
+  val KING_ATTACKS       = Array.fill(64)(0L)
+  val WHITE_PAWN_ATTACKS = Array.fill(64)(0L)
+  val BLACK_PAWN_ATTACKS = Array.fill(64)(0L)
+
+  val BETWEEN = Array.ofDim[Long](64, 64)
+  val RAYS    = Array.ofDim[Long](64, 64)
+
+  // Large overlapping attack table indexed using magic multiplication.
+  val ATTACKS = Array.fill(88772)(0L)
+
+  /** Slow attack set generation. Used only to bootstrap the attack tables.
+    */
+  def slidingAttacks(square: Int, occupied: Bitboard, deltas: Array[Int]): Bitboard =
+    var attacks = 0L
+    deltas.foreach { delta =>
+      var sq: Int = square
+      var i       = 0
+      while
+        i += 1
+        sq += delta
+        val con = (sq < 0 || 64 <= sq || distance(sq, sq - delta) > 2)
+        if (!con)
+          attacks |= 1L << sq
+
+        !(occupied.contains(sq) || con)
+      do ()
+    }
+    attacks
+
+  def initMagics(square: Int, magic: Magic, shift: Int, deltas: Array[Int]) =
+    var subset = 0L
+    while
+      val attack = slidingAttacks(square, subset, deltas)
+      val idx    = ((magic.factor * subset) >>> (64 - shift)).toInt + magic.offset
+      ATTACKS(idx) = attack
+
+      // Carry-rippler trick for enumerating subsets.
+      subset = (subset - magic.mask) & magic.mask
+
+      subset != 0
+    do ()
+
+  def initialize() =
+    (0 until 8).foreach { i =>
+      RANKS(i) = 0xffL << (i * 8)
+      FILES(i) = 0x0101010101010101L << i
+    }
+
+    val squareRange = 0 until 64
+    squareRange.foreach { sq =>
+      // println(s"$sq")
+      KNIGHT_ATTACKS(sq) = slidingAttacks(sq, Bitboard.ALL, KNIGHT_DELTAS)
+      KING_ATTACKS(sq) = slidingAttacks(sq, Bitboard.ALL, KING_DELTAS)
+      WHITE_PAWN_ATTACKS(sq) = slidingAttacks(sq, Bitboard.ALL, WHITE_PAWN_DELTAS)
+      BLACK_PAWN_ATTACKS(sq) = slidingAttacks(sq, Bitboard.ALL, BLACK_PAWN_DELTAS)
+
+      initMagics(sq, Magic.ROOK(sq), 12, ROOK_DELTAS)
+      initMagics(sq, Magic.BISHOP(sq), 9, BISHOP_DELTAS)
+    }
+
+    for
+      a <- squareRange
+      b <- squareRange
+      _ =
+        if slidingAttacks(a, 0, ROOK_DELTAS).contains(b) then
+          BETWEEN(a)(b) = slidingAttacks(a, 1L << b, ROOK_DELTAS) & slidingAttacks(b, 1L << a, ROOK_DELTAS)
+          RAYS(a)(b) =
+            (1L << a) | (1L << b) | slidingAttacks(a, 0, ROOK_DELTAS) & slidingAttacks(b, 0, ROOK_DELTAS)
+        else if slidingAttacks(a, 0, BISHOP_DELTAS).contains(b) then
+          BETWEEN(a)(b) =
+            slidingAttacks(a, 1L << b, BISHOP_DELTAS) & slidingAttacks(b, 1L << a, BISHOP_DELTAS)
+          RAYS(a)(b) =
+            (1L << a) | (1L << b) | slidingAttacks(a, 0, BISHOP_DELTAS) & slidingAttacks(b, 0, BISHOP_DELTAS)
+    yield ()
+
+  initialize()
+
+  given Conversion[Pos, Int] = _.value
```

Comment:
@ornicar Now I use this for implicit conversion between Pos and Int. Not sure it's the best way though?

If we use `opaque type Pos <: Int = Int`, we can use Pos whenever we use Int but not the other way around. But I think you have some reason why you don't like it.

Reply:
Indeed I never use this bound as it negates half the benefits of opaque types.
It's also a bad idea to use `Conversion[Pos, Int]` anywhere but in tests.

What happens if you remove it? The code will require a few more `.value` that will be erased at compilation? Maybe a few extension defs will make it more readable?

Reply:
ok, I'll try to use `.value` instead of `implicit conversion`.
But I expect there will be more than a few more `.value` because Bitboards uses a lot of array access as well as binary operators.

Also will try to refactor to reduce `.value` noise.

Reply:
Maybe it's ok to just work with primitives in the functions where the `.value` noise gets too loud.